File: client/src/app/_services/post.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { LinkPreviewData, Post } from "app/models";
import { Observable } from "rxjs";

const POST_URL = "/api/posts"

@Injectable()
export class PostService{

    constructor(private http: HttpClient){}

    getPosts(category?: string, tag?: string, followers?: string[], symbols?:string[]): Observable<Post[]> {
      let params = new HttpParams();
      if(category){
        params = params.set('category',category);
      }
      if(tag){
        params = params.set('tag',tag)
      }
      if(followers){
        const followersStr = followers.join(',')
        params = params.set('followers', followersStr)
      }
      if(symbols){
        const symbolsStr = symbols.join(',')
        params = params.set('symbols', symbolsStr)
      }
        return this.http.get<Post[]>(POST_URL,{params});
      }

    createPost(post: Post): Observable<Post> {
        return this.http.post<Post>(POST_URL, post);
      }

    getPostsByEmail(email: String): Observable<Post[]>{
      return this.http.get<Post[]>(`${POST_URL}/${email}`);
    }

    getLinkPreview(url: string): Observable<LinkPreviewData>{
      return this.http.post<LinkPreviewData>(`${POST_URL}/getMetaData`,{url});
    }
}

File: server/src/main/java/nus/iss/tfip/miniProject/services/PostService.java
package nus.iss.tfip.miniProject.services;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import nus.iss.tfip.miniProject.models.Post;
import nus.iss.tfip.miniProject.payload.response.LinkPreviewData;
import nus.iss.tfip.miniProject.repositories.jpa.PostRepository;

@Service
public class PostService {
    
    @Autowired
    private PostRepository postRepository;

    @Autowired
    private UserStatsService userStatsService;

    public List<Post> getAllPosts(String category, String tag, List<String> followers, List<String> symbols) {
        List<Post> posts = postRepository.findAll();

        // Filter by category, if specified
        if (category != null && !category.isEmpty()) {
            posts = posts.stream()
                .filter(post -> post.getCategory().equals(category))
                .collect(Collectors.toList());
        }

        // Filter by tag, if specified
        if (tag != null && !tag.isEmpty()) {
            posts = posts.stream()
                .filter(post -> post.getTag().contains(tag))
                .collect(Collectors.toList());
        }

        // Filter by followers, if specified
        if (followers != null && !followers.isEmpty()) {
            posts = posts.stream()
                .filter(post -> followers.contains(post.getAuthorEmail()))
                .collect(Collectors.toList());
        }   
        
        // Filter by symbols, if specified
        if (symbols != null && !symbols.isEmpty()) {
            posts = posts.stream()
                .filter(post -> post.convertTags().stream().anyMatch(t -> symbols.contains(t)))
                .collect(Collectors.toList());
        }  

        return posts;
    }

    public LinkPreviewData getLinkPreview(String url){
        Document document;
        try{
            document = Jsoup.connect(url).get();
            String title = document.select("meta[property=og:title]").get(0).attr("content");
            String description = document.select("meta[property=og:description]").get(0).attr("content");
            String imageURL = document.select("meta[property=og:image]").get(0).attr("content");
            return new LinkPreviewData(url, title, description, imageURL);
        } catch (IOException e){
            e.printStackTrace();
            return new LinkPreviewData();
        }
        
    }

    @Transactional
    public Post createPost(Post post) {
        Post newPost = postRepository.save(post);
        userStatsService.updateStats(post.getAuthorEmail());
        return newPost;
    }

    public List<Post> getPostsByEmail(String email){
        return postRepository.findAllByEmail(email);
    }
}
